wagtail/contrib/postgres_search/query.py

```diff
@@ -0,0 +1,125 @@
+# Originally from https://github.com/django/django/pull/8313
+# Resubmitted in https://github.com/django/django/pull/12727
+
+from django.contrib.postgres.search import SearchQueryCombinable, SearchQueryField
+from django.db.models.expressions import Expression, Value
+
+
+class LexemeCombinable(Expression):
+    BITAND = '&'
+    BITOR = '|'
+
+    def _combine(self, other, connector, reversed, node=None):
+        if not isinstance(other, LexemeCombinable):
+            raise TypeError(
+                'Lexeme can only be combined with other Lexemes, '
+                'got {}.'.format(type(other))
+            )
+        if reversed:
+            return CombinedLexeme(other, connector, self)
+        return CombinedLexeme(self, connector, other)
+
+    # On Combinable, these are not implemented to reduce confusion with Q. In
+    # this case we are actually (ab)using them to do logical combination so
+    # it's consistent with other usage in Django.
+    def bitand(self, other):
+        return self._combine(other, self.BITAND, False)
+
+    def bitor(self, other):
+        return self._combine(other, self.BITOR, False)
+
+    def __or__(self, other):
+        return self._combine(other, self.BITOR, False)
+
+    def __and__(self, other):
+        return self._combine(other, self.BITAND, False)
+
+
+class Lexeme(LexemeCombinable, Value):
+    _output_field = SearchQueryField()
+
+    def __init__(self, value, output_field=None, *, invert=False, prefix=False, weight=None):
+        self.prefix = prefix
+        self.invert = invert
+        self.weight = weight
+        super().__init__(value, output_field=output_field)
+
+    def as_sql(self, compiler, connection):
+        param = "'%s'" % self.value.replace("'", "''").replace("\\", "\\\\")
+
+        template = "%s"
+
+        label = ''
+        if self.prefix:
+            label += '*'
+        if self.weight:
+            label += self.weight
+
+        if label:
+            param = '{}:{}'.format(param, label)
+        if self.invert:
+            param = '!{}'.format(param)
+
+        return template, [param]
+
+
+class CombinedLexeme(LexemeCombinable):
+    _output_field = SearchQueryField()
+
+    def __init__(self, lhs, connector, rhs, output_field=None):
+        super().__init__(output_field=output_field)
+        self.connector = connector
+        self.lhs = lhs
+        self.rhs = rhs
+
+    def as_sql(self, compiler, connection):
+        value_params = []
+        lsql, params = compiler.compile(self.lhs)
+        value_params.extend(params)
+
+        rsql, params = compiler.compile(self.rhs)
+        value_params.extend(params)
+
+        combined_sql = '({} {} {})'.format(lsql, self.connector, rsql)
+        combined_value = combined_sql % tuple(value_params)
+        return '%s', [combined_value]
+
+
+class RawSearchQuery(SearchQueryCombinable, Expression):
```

Comment:
@kaedroho Please can you explain what this class is doing that Django's own `SearchQuery` class doesn't do? My best guess is that it's the equivalent of the two-line patch to `SearchQuery` in https://github.com/django/django/pull/12727 , but that implementation has changed so drastically between Django 3.0 and master that I can't see how one relates to the other.

Reply:
Up until Django 3.1, it wasn't possible to pass in expressions as the query string: https://github.com/django/django/commit/3baf92cf8230ad3a932986170fd07c8feae7ff2f#diff-81374e85d785546664c64c9ecb89b085